middleman_protocol/middleman_protocol/message.py

```diff
@@ -0,0 +1,387 @@
+
+    @classmethod
+    def add_frame_separator(cls, raw_frame: bytes) -> bytes:
+        """ Adds frame separator to raw frame. """
+        assert isinstance(raw_frame, bytes)
+
+        return raw_frame + FRAME_SEPARATOR
+
+    @classmethod
+    def remove_frame_separator(cls, raw_message: bytes) -> bytes:
+        """ Remove frame separator from raw frame. """
+        assert isinstance(raw_message, bytes)
+        assert FRAME_SEPARATOR in raw_message
+
+        frame_separator_index = raw_message.index(FRAME_SEPARATOR)
+
+        assert len(raw_message) == frame_separator_index + FRAME_SEPARATOR_LENGTH
+
+        return raw_message[:frame_separator_index]
+
+    @classmethod
+    def deserialize(
+        cls,
+        raw_message: bytes,
+        public_key: bytes,
+        is_frame_separator_expected: bool=False,
+        is_encoding_expected: bool=False,
+    ) -> Any:
+        """
+        Parses and validates received message in bytes.
+        Returns deserialized frame with deserialized payload.
+        """
+        assert isinstance(is_frame_separator_expected, bool)
+        assert isinstance(is_encoding_expected, bool)
+
+        # Remove frame separator from raw message if it is expected
+        if is_frame_separator_expected:
+            raw_message = cls.remove_frame_separator(raw_message)
+
+        # Reverse encoding on raw message if it is expected
+        if is_encoding_expected:
+            raw_message = cls.escape_decode_raw_message(raw_message)
+
+        # Parse frame
+        frame_format = cls.get_frame_format()
+        try:
+            frame = frame_format.parse(raw_message)
+        except StreamError as exception:
+            raise FrameInvalidMiddlemanProtocolError(f'Deserializing message failed with exception {exception}.')
+
+        raw_signed_part_of_the_frame = frame_format.signed_part_of_the_frame.build(
+            frame.signed_part_of_the_frame
+        )
+
+        # Validate
+        cls._validate_signature(raw_signed_part_of_the_frame, frame.frame_signature, public_key)
+        cls._validate_length(frame)
+        cls._validate_payload_type(frame.signed_part_of_the_frame.payload_type)
+
+        # Get class related to current payload type
+        message_class = PAYLOAD_TYPE_TO_MIDDLEMAN_MESSAGE_CLASS[
+            PayloadType[str(frame.signed_part_of_the_frame.payload_type)]
+        ]
+
+        # Deserialize payload
+        deserialized_payload = message_class._deserialize_payload(frame.signed_part_of_the_frame.payload)
+
+        # Recreate original message
+        message = message_class(
+            payload=deserialized_payload,
+            request_id=frame.signed_part_of_the_frame.request_id
+        )
+
+        return message
+
+    @classmethod
+    def escape_encode_raw_message(cls, raw_message: bytes) -> bytes:
+        """ Escapes occurrences of escape character and frame separator in raw message to avoid sending them in message. """
+        return raw_message.replace(
```

Comment:
This is not the right place for escaping. You should have a function that's the reverse of the `handle_frame_receive()` generator. That function should handle escaping and inserting separators.

Reply:
I have moved those methods to helpers, but I still think there should be possibility to use those helpers on demand in `.serialize()` for convenience. So far almost everywhere I have used `serialize()` (in 'future' pull requests too), I use it with those options.

Reply:
Just for reference: we talked about it after standup and we agreed that the only reason for it being inside `AbstractMiddlemanMessage` is convenience so we can replace it with a helper function that will be just as convenient but defined together with other code for processing the byte stream.